### ISSUE TITLE
[torch.compile] VonMises.sample() is not capturable with fullgraph=True because its rejection-sampling loop uses data-dependent control flow

### ISSUE THREAD

**Issue by PiesAvalon** [pytorch/pytorch#194764]
### 🐛 Describe the bug

`torch.distributions.VonMises.sample()` cannot be captured by
`torch.compile(fullgraph=True)` even when the `VonMises` distribution is
constructed outside the compiled function.

With that setup, the failure is not the generic decorated-method
`SourcelessBuilder` case. Dynamo reaches the implementation of
`VonMises.sample()` and rejects its rejection-sampling loop with the expected
but unusable-for-fullgraph error:

```text
Unsupported: Data-dependent branching
Explanation: Detected data-dependent branching
Developer debug context: attempted to jump with TensorVariable()
```

The current implementation uses tensor-dependent loop and branch conditions
(`while not done.all()` and `if accept.any()`). This is a separate,
VonMises-specific compile-support limitation: it remains after the distribution
object and its decorated method have been isolated outside the compiled region.

The failure occurs with `backend="eager"`, `backend="aot_eager"`, and
Inductor, so it occurs in Dynamo before backend lowering. With the default
`fullgraph=False`, Dynamo graph-breaks around the unsupported code and eager
fallback returns a valid sample.

This is a compile-support issue for a public distribution sampling API, not a
claim of numerical incorrectness. The eager result and the graph-break fallback
remain numerically valid; the missing behavior is full-graph capture.

## Minimal reproduction

```python
import torch
from torch.distributions import VonMises


distribution = VonMises(
    torch.tensor([0.3]),
    torch.tensor([2.0]),
)


def fn():
    return distribution.sample()


torch.manual_seed(123)
print("eager:", fn())

torch.manual_seed(123)
compiled = torch.compile(fn, backend="eager", fullgraph=True)
print("compiled:", compiled())
```

## Actual behavior

```text
eager: tensor([...])
torch._dynamo.exc.Unsupported: Data-dependent branching
  Explanation: Detected data-dependent branching (e.g. `if my_tensor.sum() > 0:`).
  Hint: Use `torch.cond` to express dynamic control flow.

  Developer debug context: attempted to jump with TensorVariable()

from user code:
  ...
  return distribution.sample()
```

The exact sampled value is not important; it varies with the random seed and
PyTorch version. The important distinction is the exception type and that it is
raised while capturing `_rejection_sample`, before backend lowering.

## Expected behavior

The public `VonMises.sample()` API should either provide a compiler-compatible
implementation for `fullgraph=True`, or expose a documented compile-compatible
path for sampling. At minimum, the public API should have a clear support story
for `torch.compile`; currently the only working mode is graph-break fallback.

## Controls and ablation

1. Calling `distribution.sample()` eagerly succeeds and returns a tensor with
   the expected shape.
2. Compiling the same `fn` with `fullgraph=False` succeeds through eager
   fallback for the tested backends.
3. Constructing the distribution outside `fn` removes the generic
   `SourcelessBuilder` failure and consistently exposes the
   `Data-dependent branching` limitation in `_rejection_sample`.
4. The same full-graph failure is observed with `backend="eager"`,
   `backend="aot_eager"`, and Inductor.

The rejection-sampling loop is the essential trigger. This report does not rely
on the separate generic failure involving an in-graph object whose method is
decorated with `@torch.no_grad()`.

## Reproducibility

The full-graph failure reproduced in 3/3 independent processes on the stable
environment listed below. The `fullgraph=False` path returned a valid sample
after graph-break/eager fallback.

## Related issue / duplicate search

This report is intentionally scoped to the `VonMises.sample()` implementation
and its tensor-dependent rejection-sampling control flow. A smaller standalone
report should track generic `SourcelessBuilder` handling of grad-mode-decorated
methods; that generic issue is not required for this reproducer.

### Versions

```
Collecting environment information...
PyTorch version: 2.12.1+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 591.86
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               28
On-line CPU(s) list:                  0-27
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i7-14700K
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   14
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             6835.20
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni vnmi umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            672 KiB (14 instances)
L1i cache:                            448 KiB (14 instances)
L2 cache:                             28 MiB (14 instances)
L3 cache:                             33 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-27
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.9.2.10
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.9.86
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu12==9.24.0.43
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] onnxruntime==1.24.4
[pip3] torch==2.12.1
[pip3] torchvision==0.27.1
[pip3] triton==3.7.1
[conda] nvidia-cublas               13.1.1.3         pypi_0              pypi
[conda] nvidia-cuda-cupti           13.0.85          pypi_0              pypi
[conda] nvidia-cuda-nvrtc           13.0.88          pypi_0              pypi
[conda] nvidia-cuda-runtime         13.0.88          pypi_0              pypi
[conda] nvidia-cudnn-cu13           9.20.0.48        pypi_0              pypi
[conda] nvidia-cufft                12.0.0.61        pypi_0              pypi
[conda] nvidia-curand               10.4.0.35        pypi_0              pypi
[conda] nvidia-cusolver             12.0.4.66        pypi_0              pypi
[conda] nvidia-cusparse             12.6.3.3         pypi_0              pypi
[conda] nvidia-cusparselt-cu13      0.8.1            pypi_0              pypi
[conda] nvidia-nccl-cu13            2.29.7           pypi_0              pypi
[conda] nvidia-nvjitlink            13.0.88          pypi_0              pypi
[conda] nvidia-nvtx                 13.0.85          pypi_0              pypi
[conda] torch                       2.12.1           pypi_0              pypi
[conda] torchvision                 0.27.1           pypi_0              pypi
[conda] triton                      3.7.1            pypi_0              pypi
```

cc @fritzo @neerajprad @alicanb @nikitaved @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98
